### PR TITLE
Porting socket context changes to the CLI

### PR DESCRIPTION
**Pull request by amit-jain9**
Porting socket context changes to the CLI:

1. Used Attempt Entry wrapper structure to store the socket context to be used while socket connections.
2. Code changes to free the context with attempt entry on cli disconnect.

### CODE REVIEW

**Review by johnmeneghini — APPROVED** [review]
These changes look good to me.

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c [hunk 1]
```diff
@@ -78,20 +78,28 @@ NvmeOfCliDeleteMapEntries (
...
+  LIST_ENTRY               *NextEntryProcessed = NULL;
   NVMEOF_CLI_CTRL_MAPPING  *MappingList;
   BOOLEAN                  Flag = FALSE;
+  VOID                     *Private = NULL;
 
-  NET_LIST_FOR_EACH (Entry, &gCliCtrlMap->CliCtrlrList) {
+  NET_LIST_FOR_EACH_SAFE (Entry, NextEntryProcessed, &gCliCtrlMap->CliCtrlrList) {
     MappingList = NET_LIST_USER_STRUCT (Entry, NVMEOF_CLI_CTRL_MAPPING, CliCtrlrList);
     if (MappingList->Ctrlr == Ctrlr) {
+      if (MappingList->Private != Private) {
```
Comment (mwilck) [inline NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:89]: Please explain this comparison.

You seem to want to make sure that you don't free a MappingList->Privat pointer more than once. But it isn't safe to do this this way. What if the MappingList->Private pointer has values A, B, and A again while you traverse the list?

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c [hunk 2]
```diff
@@ -919,6 +931,36 @@ NvmeOfCliProbeCallback (
       CopyMem (Opts->hostnqn, ProbeconnectData->Hostnqn, sizeof (ProbeconnectData->Hostnqn));
     }
   }
+
+  // Fill socket context
+  Private     = (NVMEOF_DRIVER_DATA*)CallbackCtx;
+  AttemptData = &Private->Attempt->Data;
```
Comment (mwilck) [inline NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:937]: Can you assume that Private->Attempt is non-NULL and valid?

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c [hunk 3]
```diff
@@ -919,6 +931,36 @@ NvmeOfCliProbeCallback (
...
+  Context     = &Private->Attempt->SocketContext;
+
+  Context->Controller = Private->Controller;
+  Context->IsIp6      = AttemptData->SubsysConfigData.NvmeofIpMode == IP_MODE_IP6;
+
+  if (!Context->IsIp6) {
```
Comment (mwilck) [inline NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:943]: And what if it _is_ IPv6 ?

**Review by mwilck — COMMENTED** [review]
Some comments in the code.

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c (hunk 1 above)
Comment (amit-jain9) [reply NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:89]: Yes, MappingList entries are created one per each namespace of a controller in NvmeOfAttachCliCallback(). Ctrlr and Private are common across all the entries saved during one execution of NvmeOfAttachCliCallback. All such entries in the list will be one after the other i.e., values like A, B and A again are not possible. Private can differ in subsequent entries of the list only if they are namespaces from another connect command using CLI. Connecting to the same controller again is not allowed using CLI.

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c (hunk 2 above)
Comment (amit-jain9) [reply NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:937]: Yes, Private always contains attempt data for which the Probe callback is invoked. Attempt data is populated to Private before invoking spdk_nvme_probe, while processing a connect command from CLI. Same Private is passed as callback parameter to Probe and Attach callback.

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c (hunk 3 above)
Comment (amit-jain9) [reply NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:943]: We don't use Host specific details for creating/connecting to the sockets, while using IPv6. 
https://github.com/timberland-sig/edk2/blob/05081056a30e34019da707fe247731f9f58c0743/NetworkPkg/Library/DxeSpdkLib/SpdkShim/edk_sock.c#L171

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c (hunk 1 above)
Comment (mwilck) [reply NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:89]: > values like A, B and A again are not possible

It's hard for me to verify this, so I believe you. But I suggest you add a comment explaining the intention of the code, and why it works the way it's intended to work.

File: NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c (hunk 3 above)
Comment (mwilck) [reply NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:943]: I think this deserves a comment in the code, too.

**Review by mwilck — APPROVED** [review]
All that's left is that I asked for some comments. It's fine, approving.

**Comment by johnmeneghini** [comment]
I'd like to approve this merge request.

I've tested this patch together with all of the current EDK2 changes and the latest timberland dracut changes from https://github.com/dracutdevs/dracut/pull/2184 and everything works in the simple case. 

I used the [Developer Build](https://github.com/timberland-sig/rh-linux-poc#developer-build) process on the [rh-linux-poc](https://github.com/timberland-sig/rh-linux-poc) to compile, build, and test the EDK2 branches from pulls requests #18 #17 and #5.  The combined branch I tested can be seen at: https://github.com/timberland-sig/edk2/tree/populate_nbft_fail_case-johnm

Please merge this pull request.

**Pull request merged**